### PR TITLE
Fix division by zero in clDice loss harmonic mean

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
## Summary

- Fix division by zero in `SoftclDiceLoss` and `SoftDiceclDiceLoss` when computing the harmonic mean of topology precision and sensitivity
- Add a small epsilon (`1e-7`) to the denominator `(tprec + tsens)` to prevent `NaN` when both values are zero
- Add test cases for zero-input and non-overlapping edge cases with `smooth=0`

## Details

The clDice loss computes `cl_dice = 1.0 - 2.0 * (tprec * tsens) / (tprec + tsens)`. When both `tprec` and `tsens` are zero (e.g., empty inputs, non-overlapping predictions/targets, or `smooth=0`), this results in `0/0 = NaN`, which propagates through the loss and crashes training.

While the default `smooth=1.0` prevents `tprec` and `tsens` from being exactly zero in most cases, setting `smooth=0` (a valid configuration) exposes this bug whenever skeleton overlap is zero. The fix adds `1e-7` to the harmonic mean denominator, which:

- Has negligible impact on normal computation (tprec, tsens are bounded in [0, 1])
- Returns `cl_dice = 1.0` (maximum loss) when both precision and sensitivity are zero, which is the correct semantic result
- Is consistent with epsilon-based denominator guards used elsewhere in MONAI (e.g., `smooth_dr` in `DiceLoss`)

## Test plan

- [x] Existing `test_cldice_loss.py` tests still pass (perfect overlap cases)
- [x] New `test_zero_input_no_nan`: verifies zero-valued inputs with `smooth=0` do not produce NaN
- [x] New `test_no_overlap_no_nan`: verifies non-overlapping predictions/targets with `smooth=0` do not produce NaN

### CODE REVIEW

**Comment by ericspod** [comment]
Hi @Mr-Neutr0n thanks for this change, there should be a value to prevent divide-by-zero but I would suggest following the [pattern found with the other Dice losses](https://github.com/Project-MONAI/MONAI/blob/342bd7a2de175b8222644598d24499239d3273a8/monai/losses/dice.py#L67). I would add a `smooth_dr` constructor argument as a new last argument with a value of `1e-7` and use that value in your forward methods.

**Comment by Mr-Neutr0n** [comment]
Thanks @ericspod — the `smooth_dr` pattern makes much more sense. Restructured the fix to follow it:

- Added `smooth_dr: float = 1e-7` as a constructor argument to `SoftclDiceLoss`, `SoftDiceclDiceLoss`, and the `soft_dice` helper (matching `monai.losses.dice.DiceLoss`).
- Used `self.smooth_dr` in the `tprec`, `tsens`, `soft_dice`, and harmonic-mean denominators — not just the harmonic mean.

The original commit only patched the harmonic-mean denominator, but the same 0/0 bug existed in the individual `tprec` and `tsens` ratios and in `soft_dice`. With `smooth=0` and an all-zero skeleton, each ratio resolved to `0/0 = NaN` before the harmonic mean was even computed.

Verified locally — all 5 tests in `test_cldice_loss.py` pass, including the new `test_zero_input_no_nan` and `test_no_overlap_no_nan` (which were failing on the previous commit). The `1e-7` epsilon has negligible impact on normal computation since `tprec` and `tsens` are bounded in `[0, 1]`.

Latest push: `746afac`.

**Comment by ericspod** [comment]
Hi @Mr-Neutr0n please have a look at the conflicts that need resolving, There's been changes to the section of code you wanted to update, unfortunately you'll have to untangle where your changes should go.

**Comment by Mr-Neutr0n** [comment]
Hi @ericspod — I rebased onto the current dev branch and resolved the conflict.

While resolving it I saw that the core guard you asked for is already present on dev (added in #8703): the precision/sensitivity denominators use smooth_dr and the harmonic mean uses smooth, with smooth validated as positive. So the source-level division-by-zero issue is already covered upstream.

I turned this PR into the missing regression coverage instead:
- Added zero-input finite-output tests for both SoftclDiceLoss and SoftDiceclDiceLoss.
- Added non-default smooth_dr tests to verify the parameter is wired through and actually changes the result.
- Clarified the SoftclDiceLoss smooth_dr docstring.

The branch is now clean on top of dev. Let me know if you'd rather I close this and open a fresh test-only PR.

**Comment by ericspod** [comment]
Hi @Mr-Neutr0n I guess other changes got to this fix first, sorry about that. I'm fine with keeping this PR and merging shortly, if you could add the tests as Coderabbit suggests we'd be good to go.

File: tests/losses/test_cldice_loss.py [hunk 1]
```diff
@@ -114,6 +114,22 @@ def test_invalid_iter_value(self):
...
+        input_tensor = torch.zeros((1, 2, 4, 4))
+        target = torch.zeros((1, 2, 4, 4))
+        loss_a = SoftclDiceLoss(smooth=1e-7, smooth_dr=1e-3)
+        loss_b = SoftclDiceLoss(smooth=1e-7, smooth_dr=1e-5)
+        result_a = loss_a(input_tensor, target)
+        result_b = loss_b(input_tensor, target)
+        self.assertTrue(torch.isfinite(result_a).all())
+        self.assertTrue(torch.isfinite(result_b).all())
+        self.assertNotAlmostEqual(result_a.item(), result_b.item(), places=5)
+
```
[the remark this answers is not in the document]
Comment (ericspod) [reply tests/losses/test_cldice_loss.py:132]: @coderabbitai write a separate comment with a "suggestion" section adding these new tests, and which can be merged directly through Github.

**Comment by Mr-Neutr0n** [comment]
Hi @ericspod — added the non-overlap finiteness tests for both SoftclDiceLoss and SoftDiceclDiceLoss as suggested. The branch is rebased on dev and should be ready for another look.

**Review by ericspod — APPROVED** [review]
@Mr-Neutr0n looks good now thanks, I can pass your DCO for you if it's okay with you.

**Comment by Mr-Neutr0n** [comment]
Thanks for the review, @ericspod. Yes, please go ahead and pass the DCO for me — the commit messages are already signed-off. Let me know if anything else is needed.
